libmscore/types.h

```diff
@@ -423,9 +423,11 @@ enum class Tid {
 
 //---------------------------------------------------------
 //   Align
+///  \since 3.2.4
 //---------------------------------------------------------
 
-enum class Align : char {
+enum class Align {
```

Comment:
Was removing `: char` necessary here for some reason? It looks like a totally harmless change, but what was the reason to do it?

Reply:
All the enums that I encountered that were moved to types.h for v3.0 seemed to have that char subclass removed. Near as I could tell it provides nothing except generating the question:: Why is it there? 

It appeared to be a code barnacle so I removed it per the observed types.h pattern. 

-Dale

Reply:
Those back in the day got added (by me) to save RAM. Reduced the memory footprint of MuseScore by quite a bit then. But this may come at a performance penalty, so got reverted for some.

Reply:
RAM probably isn't an issue these days. :)

The only other reason to do something like this is to control serializer packing/unpacking if binary serialization is used. 

Nothing like that is happening here that I can see. 

Allowing natural data sizes lets the C++ compiler to do what it should do best, optimize.

With the high abstraction level of current C++ code it almost looks like an anachronism.

-Dale

Reply:
It still is an issue an Netbook, Chromebook, Raspeberry Pi, etc ;-)

Reply:
> Allowing natural data sizes lets the C++ compiler to do what it should do best, optimize.

According to [cppreference](https://en.cppreference.com/w/cpp/language/enum#Scoped_enumerations), a *scoped* enumeration (which is what we have here) without an underlying type specified has `int` as its underlying type. This doesn't seem to allow a room for optimizations.

In any case, while I don't know how large is an impact of this optimization and whether it does anything measurable for RAM usage on mobile or any other devices, let's better keep different changes separately. This pull request is about exposing an enumeration to plugins API, and changing the underlying type of that enumeration seems to be unrelated to this topic at all. So I believe it would be preferable to consider such changes within a separate pull request and leave this pull request with plugins-related changes only.

But the question itself is interesting, I don't have a good answer on it for sure.

Reply:
>I don't think any optimizer can auto-shrink those enums

They don't need to. The memory consumption of a single in-memory jpg or png image would swamp the total enum (for instance) storage requirement even on those platforms. i.e., The additional storage (over the forced storage class of char) would be imperceptible. 

Note that the minimum RAM you'll find on those targets is at least 4GB.

We always preached code clarity over optimization because most times programmers miss-guess what actually will provide actual optimization. 

It fact the type of optimization can be different for say MCU based processors (having no CPU cache) and app processors like ARM and Intel which do. For instance, the MCU optimizer would unroll loops and the app processor optimizer wouldn't because loops are cache friendly.

Just an old guy's observation and opinion. Prior to retiring I worked in the embedded processor space and we always whined about available footprint but we need to be realistic.

Down to this situation...

Should it be restored or not? My vote is clear but I don't call the shots here.

Dale

Reply:
A single such enum certainly is not an issue at all, 3 saved bytes are peanuts. But once they get used in arrays/vectors this sums up and in some cases by quite a bit..
But probably not worth the effort.

Reply:
Just a follow-up to the optimization discussion....

Another problem with forcing an enum down to byte-sized is that when you store them in struct/classes the compiler may force gaps in memory if an int aligned value (for instance) follows a byte aligned value. It does this so it can access the int values on it's naturally aligned address boundary (divisible by 4). 

In fact, on an ARM and many RISC platforms you will get a memory access fault if you don't access on data sized aligned boundaries. What you may not see is that in the background Linux (or other OS) will catch the fault and simulate an unaligned access in the kernel just so the program can be happy. This is not cheap overhead! Sticking with natural alignment avoids this.

Of course you can force struct/class byte aligned packing to get smaller data RAM footprints but that can slow down the program with unaligned access simulation. i.e., you trade code and speed for RAM.

Intel is more forgiving but it still can pay a price if the memory isn't on a cache line.

Next point of concern is the stack and automatic variables. 

On many processors allocating a byte sized item may still cause the allocation of more unused memory to align the data for the reason given earlier.

All this is typically hidden from most programmers. If you work with embedded systems however it needs to be considered.

In the end optimization is best left to the compiler with the programmer stepping in when a special situation is encountered.

-Dale

Reply:
Of course, having `char` as the underlying type alone is not sufficient to have memory usage actually lower, it is also necessary to design data types considering memory alignment issues (for example, ordering member variables in classes and structures from largest to smallest ones to avoid "gaps" introduced by a compiler due to alignment).

But, although it is not absolutely critical for this particular PR, my preference would be to do such changes separately from from pull requests adding some functionality (or at least do them in separate commits). Git history clarity is not less important than code clarity.

Also, if this change is still going to be made, just removing `: char` is not all what should be done, `static_cast<char>` statements for operators declared for `Align` enumeration should also better be changed to use the new underlying type:
https://github.com/musescore/MuseScore/blob/3371d9bcb2491fd66a5dd5d7e30edd54ca709883/libmscore/types.h#L443-L451

Reply:
>Also, if this change is still going to be made, just removing : char is not all what should be done, static_cast<char> statements for operators declared for Align enumeration should also better be changed to use the new underlying type:

Very good pick up. I missed those. They would still work but if the enum values got larger that a byte data would be ignored. Although I'm not sure the static_casts were even needed in the first place.

Since there is seems to be a lot of concern about this I'll restore the char and leave its consideration for another commit.

-Dale